tests/prefetch_related/tests.py

```diff
@@ -256,6 +258,85 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
         self.assertWhereContains(sql, self.author1.id)
 
 
+class LargePrefetchSetTests(TestCase):
+    TEST_SIZE = 1010
+
+    @classmethod
+    def setUpTestData(cls):
+        # Create a large set of books and authors
+        cls.book1 = Book.objects.create(title='Poems')
+        cls.book2 = Book.objects.create(title='Poems II')
+
+        authors = [Author(name='Charlotte {}'.format(i), first_book=cls.book1)
+                   for i in range(cls.TEST_SIZE)]
+        Author.objects.bulk_create(authors)
+        # Re-fetch the authors as .bulk_create() will not populate their .pk attribute
+        # on SQLite
+        cls.authors = list(Author.objects.all())
+
+        cls.reader = Reader.objects.create(name='Amy')
+
+        # We add related objects to the first and the last authors in the list so we
+        # can be sure of the result sets being properly merged after the bulk processing
+        cls.authors[0].books.add(cls.book1)
+        cls.authors[-1].books.add(cls.book2)
+        cls.reader.books_read.add(cls.book1)
+        cls.reader.books_read.add(cls.book2)
+
+    def test_number_of_queries(self):
+        if connection.features.max_query_params is not None:
+            # The current implementation uses a safety margin of 25 to
+            # cope with additional variables in the query.
+            batch_size = connection.features.max_query_params - 25
+            batches = int(ceil(float(len(self.authors)) / batch_size))
+        else:
+            batches = 1
+
+        #   1 query for looking up authors
+        # + N queries for looking up books
+        with self.assertNumQueries(1 + batches):
+            result = list(Author.objects.prefetch_related('books'))
+
+        with self.assertNumQueries(0):
+            self.assertEqual(self.TEST_SIZE, len(result))
+            self.assertEqual(1, len(result[0].books.all()))
+            self.assertEqual(1, len(result[-1].books.all()))
+
+    def test_recursive_lookups(self):
```

Comment:
I saw the term for this being recursive somewhere, but related might be even better